pkg/ccl/spanconfigccl/spanconfigreconcilerccl/spanconfigreconcilerccl_test: TestDataDriven failed

pkg/ccl/spanconfigccl/spanconfigreconcilerccl/spanconfigreconcilerccl_test.TestDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20547311?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20547311?buildTab=artifacts#/) on release-24.1.25-rc @ [e7f0180a93e8fa2bd3b9b362e2688557dc85da73](https://github.com/cockroachdb/cockroach/commits/e7f0180a93e8fa2bd3b9b362e2688557dc85da73):


```
        /Tenant/10/Table/4{7-8}                    ignore_strict_gc=true rangefeed_enabled=true exclude_data_from_backup=true
        /Tenant/10/Table/4{8-9}                    database system (tenant)
        /Tenant/10/Table/5{0-1}                    database system (tenant)
        /Tenant/10/Table/5{1-2}                    database system (tenant)
        /Tenant/10/Table/5{2-3}                    database system (tenant)
        /Tenant/10/Table/5{3-4}                    database system (tenant)
        /Tenant/10/Table/5{4-5}                    database system (tenant)
        /Tenant/10/Table/5{5-6}                    database system (tenant)
        /Tenant/10/Table/5{6-7}                    database system (tenant)
        /Tenant/10/Table/5{7-8}                    database system (tenant)
        /Tenant/10/Table/5{8-9}                    database system (tenant)
        /Tenant/10/Table/{59-60}                   database system (tenant)
        /Tenant/10/Table/6{0-1}                    database system (tenant)
        /Tenant/10/Table/6{1-2}                    database system (tenant)
        /Tenant/10/Table/6{2-3}                    database system (tenant)
        /Tenant/10/Table/6{3-4}                    database system (tenant)
        /Tenant/10/Table/6{4-5}                    database system (tenant)
        /Tenant/10/Table/6{5-6}                    database system (tenant)
        /Tenant/10/Table/6{6-7}                    database system (tenant)
        /Tenant/11{-\x00}                          database system (tenant)
        /Tenant/12{-\x00}                          database system (tenant)
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/15068/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/spanconfigccl/spanconfigreconcilerccl/spanconfigreconcilerccl_test_/spanconfigreconcilerccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/basic:195:
        exec-sql [1 args]
        CREATE DATABASE db;
        CREATE TABLE db.t1();
        CREATE TABLE db.t2();
        CREATE SCHEMA db.sc;
        CREATE TYPE typ AS ENUM();
        CREATE VIEW v AS SELECT 1;
        CREATE SEQUENCE db.seq;
        CREATE MATERIALIZED VIEW mv AS SELECT 1;
        ----
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/15068/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/ccl/spanconfigccl/spanconfigreconcilerccl/spanconfigreconcilerccl_test_/spanconfigreconcilerccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/basic:206:
         
        expected:
        upsert /Tenant/10/Table/10{6-7}            range default
        upsert /Tenant/10/Table/10{7-8}            range default
        upsert /Tenant/10/Table/11{2-3}            range default
        upsert /Tenant/10/Table/11{3-4}            range default
        
        found:
        upsert /Tenant/10/Table/11{6-7}            range default
        upsert /Tenant/10/Table/11{7-8}            range default
        upsert /Tenant/10/Table/12{2-3}            range default
        upsert /Tenant/10/Table/12{3-4}            range default
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/bb5423f08bb702396d5348bf5fd63a5d/logTestDataDriven_multitenant_basic2540934475
        --- FAIL: TestDataDriven/multitenant/basic (13.07s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-55006

--- a/pkg/spanconfig/spanconfigtestutils/spanconfigtestcluster/cluster.go
+++ b/pkg/spanconfig/spanconfigtestutils/spanconfigtestcluster/cluster.go
@@ -71,6 +71,9 @@ func (h *Handle) InitializeTenant(ctx context.Context, tenID roachpb.TenantID) *
 				SpanConfig:     h.scKnobs,
 				GCJob:          &tenantGCJobKnobs,
 				UpgradeManager: serverUpgradeKnobs,
+				SQLExecutor: &sql.ExecutorTestingKnobs{
+					UseTransactionalDescIDGenerator: true,
+				},
 			},
 		}
 		var err error

--- a/pkg/sql/planner.go
+++ b/pkg/sql/planner.go
@@ -24,6 +24,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catsessiondata"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descidgen"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/lease"
@@ -462,6 +463,9 @@ func newInternalPlanner(
 	p.extendedEvalCtx.NodeID = execCfg.NodeInfo.NodeID
 	p.extendedEvalCtx.Locality = execCfg.Locality
 	p.extendedEvalCtx.DescIDGenerator = execCfg.DescIDGenerator
+	if execCfg.TestingKnobs.UseTransactionalDescIDGenerator && txn != nil {
+		p.extendedEvalCtx.DescIDGenerator = descidgen.NewTransactionalGenerator(execCfg.Settings, execCfg.Codec, txn)
+	}
 
 	p.sessionDataMutatorIterator = smi
 

